File: 2024/day1.py
from collections import defaultdict

with open("day1.in", "r") as infile:
    lines = infile.readlines()

    lst1 = []
    lst2 = []
    for line in lines:
        a, b = map(int, line.split())
        lst1.append(a)
        lst2.append(b)

    lst1.sort()
    lst2.sort()
    res = sum(abs(a - b) for a, b in zip(lst1, lst2))
    print(res)

with open("day1.in", "r") as infile:
    lines = infile.readlines()

    lst = []
    counts: dict[int, int] = defaultdict(lambda: 0)
    for line in lines:
        a, b = map(int, line.split())
        lst.append(a)
        counts[b] += 1

    res = sum(a * counts[a] for a in lst)
    print(res)

File: advent_rust/src/year_2024/day_1.rs
use std::{fs::read_to_string, iter::zip};

fn read_lines(filename: &str) -> Vec<String> {
    let mut result = Vec::new();

    for line in read_to_string(filename).unwrap().lines() {
        result.push(line.to_string())
    }

    result
}

pub fn part_1() -> String {
    let lines = read_lines("./src/year_2024/day_1.in");

    let mut left = vec![];
    let mut right = vec![];
    for line in lines {
        let mut parts = line.split_whitespace();
        left.push(parts.next().unwrap().parse::<i32>().unwrap());
        right.push(parts.next().unwrap().parse::<i32>().unwrap());
    }

    left.sort();
    right.sort();

    let result: i32 = zip(left, right).map(|(l, r)| (l - r).abs()).sum();

    result.to_string()
}
